engine/src/main/java/io/zeebe/engine/processing/message/MessageSubscriptionCreateProcessor.java

```diff
@@ -65,29 +70,20 @@ public void processRecord(
       return;
     }
 
-    handleNewSubscription(record, streamWriter, sideEffect);
+    handleNewSubscription(sideEffect);
   }
 
-  private void handleNewSubscription(
-      final TypedRecord<MessageSubscriptionRecord> record,
-      final TypedStreamWriter streamWriter,
-      final Consumer<SideEffectProducer> sideEffect) {
-    final MessageSubscription subscription =
-        new MessageSubscription(
-            subscriptionRecord.getWorkflowInstanceKey(),
-            subscriptionRecord.getElementInstanceKey(),
-            subscriptionRecord.getBpmnProcessIdBuffer(),
-            subscriptionRecord.getMessageNameBuffer(),
-            subscriptionRecord.getCorrelationKeyBuffer(),
-            subscriptionRecord.shouldCloseOnCorrelate());
+  private void handleNewSubscription(final Consumer<SideEffectProducer> sideEffect) {
 
-    sideEffect.accept(this::sendAcknowledgeCommand);
+    // TODO (saig0): the subscription should have a key (#2805)
+    stateWriter.appendFollowUpEvent(-1L, MessageSubscriptionIntent.CREATED, subscriptionRecord);
 
-    subscriptionState.put(subscription);
-    messageCorrelator.correlateNextMessage(subscription, subscriptionRecord, sideEffect);
+    final var isMessageCorrelated =
+        messageCorrelator.correlateNextMessage(subscriptionRecord, sideEffect);
 
-    streamWriter.appendFollowUpEvent(
-        record.getKey(), MessageSubscriptionIntent.OPENED, subscriptionRecord);
+    if (!isMessageCorrelated) {
```

Comment:
Is this right? We send the ack when not correlated?

Reply:
Yes. The processor acknowledges the command by sending a command `WorkflowInstanceSubscription:Open`. If there are messages buffered that can be correlated then it correlates a message by sending a command `WorkflowInstanceSubscription:Correlate` instead.